### PR TITLE
feat: add canonical tool

### PR DESCRIPTION
**Pull request by tbpg**
This tool, `canonical`, adds canonical links to HTML pages in tarballs in buckets. It writes the result to another bucket where content can be copied from later.

Runs are idempotent; processing the same tarball more than once won't lead to duplicate canonical links.

I experimented with a couple different values for `semaphore.NewWeighted` and 500 worked well on my machine. I also tried using https://github.com/klauspost/pgzip, but it didn't lead to any improvements because the tarballs are generally less than 1MB.

The tool is CPU bound. From profiling, most of the CPU is spent on gzip compression. I made a couple optimizations based on profiling, like avoiding regex and a couple unnecessary string allocations, but gzip really is the main issue.

Example usage:

```
$ go build
$ ./canonical -f -inbucket my-bucket-1 -outbucket my-bucket-2 -cpuprofile=cpu.prof -memprofile=mem.mprof
Canonicalizing... 1747444 files |  15200 / 15200 [----------------------------------------] 100.00% 7m50s
````

The `canonical` tool hits a Cloud Function for every HTML file. So, the above run hit ~3k QPS on that function. (Another run was ~30 seconds faster.)

Given this takes less than 10 minutes, it's not really worth optimizing too much more.

In the future, we can repurpose this to redirect pages instead of adding canonical links to them. We can also consider adding some link checking to avoid adding canonical links or redirects to 404s.

cc @jskeet

### CODE REVIEW

File: canonical/main.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+	return name
+}
+
+// canonicalLink returns the canonical link for the given package and page.
+func canonicalLink(ctx context.Context, pkg, page string) (string, error) {
+	q := url.Values{
+		"package": []string{pkg},
+		"page":    []string{page},
+	}
+	u, err := url.Parse("https://us-central1-jonskeet-integration-tests.cloudfunctions.net/canonicalize-link")
```
Comment (tritone) [inline canonical/main.go:334]: Make this a constant as well?

File: canonical/main.go [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		panic(err)
+	}
+	return req
+}
+
+// objects returns a map of object names in the bucket.
+func objects(ctx context.Context, bucket *storage.BucketHandle) (map[string]bool, error) {
+	objs := map[string]bool{}
+	it := bucket.Objects(ctx, &storage.Query{Prefix: dotnetPrefix})
```
Comment (tritone) [inline canonical/main.go:365]: If there are many objects and you're only getting the names, you can speed this up by using `ProjectionNoACL` and/or `SetAttrSelection` https://pkg.go.dev/cloud.google.com/go/storage#Query

File: canonical/main.go [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+// parsePkg gets the package name from the tarball name.
+func parsePkg(tarballName string) string {
+	name := tarballName[len(dotnetPrefix):] // Trim docfx- prefix.
+	name = name[:strings.Index(name, "-")]  // Discard after first remaining -.
+	return name
+}
+
+// canonicalLink returns the canonical link for the given package and page.
+func canonicalLink(ctx context.Context, pkg, page string) (string, error) {
```
Comment (tritone) [inline canonical/main.go:329]: Just out of curiosity, why do we need a whole separate service to provide these links?

Comment (jskeet) [reply canonical/main.go:329]: It's relatively complex logic, which we need in the .NET side of things as well (for generating canonical links in googleapis.dev in new packages, and for setting redirects in github.com). A Cloud Function is a really simple way of sharing this logic.

File: canonical/main.go [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return err
+	}
+
+	existingOutputs, err := objects(ctx, outBucket)
+	if err != nil {
+		return err
+	}
+
+	inputs := []string{}
```
Comment (codyoss) [inline canonical/main.go:145]:
```suggestions
	var inputs []string
```

nit: https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices

File: canonical/main.go [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+	for name := range allInputs {
+		if processEverything || !existingOutputs[name] {
+			inputs = append(inputs, name)
+		}
+	}
+
+	if len(inputs) == 0 {
+		fmt.Println("Nothing to do. Add -f flag?")
```
Comment (codyoss) [inline canonical/main.go:153]:
```suggestion
		log.Println("Nothing to do. Add -f flag?")
```

File: canonical/main.go [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+			log.Println(err)
+			break
+		}
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			bar.Finish()
+			log.Println(err)
+			break
+		}
+		releaseOnce := &sync.Once{}
```
Comment (codyoss) [inline canonical/main.go:174]: Why is this needed?

File: canonical/main.go [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+	}
+
+	bar := pbTemplate.Start(len(inputs)).Set("prefix", "Canonicalizing... ").SetRefreshRate(time.Second)
+
+	g, ctx := errgroup.WithContext(ctx)
+	sem := semaphore.NewWeighted(500)
+
+	for _, name := range inputs {
+		if err := ctx.Err(); ctx.Err() != nil {
+			bar.Finish()
```
Comment (codyoss) [inline canonical/main.go:164]: Can this just be defered once after bar is created?

File: canonical/main.go [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+// The progress bar is updated with the number of files processed as a prefix.
+func processTarball(ctx context.Context, inBucket, outBucket *storage.BucketHandle, name string, bar *pb.ProgressBar) error {
+	pkg := parsePkg(name)
+
+	obj := inBucket.Object(name)
+
+	// Storage reader -> un-gzip -> un-tar
+	// -> process
+	// -> tar -> gzip -> Storage writer.
+	r, err := obj.NewReader(ctx)
```
Comment (codyoss) [inline canonical/main.go:207]: I believe that Close must be called on this resource.

File: canonical/main.go [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to read from gs://%s/%s: %v", obj.BucketName(), name, err)
+	}
+
+	gr, err := gzip.NewReader(r)
```
Comment (codyoss) [inline canonical/main.go:212]: Same as above, please defer a Close.

File: canonical/main.go [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	tr := tar.NewReader(gr)
+
+	outObj := outBucket.Object(name)
+	ow := outObj.NewWriter(ctx)
```
Comment (codyoss) [inline canonical/main.go:219]: In addition to the explicit Closes later should these be defered as well for the error cases where the loop returns?

File: canonical/main.go [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+
+		// Update the header to have the right size. Otherwise, leave hdr unchanged.
+		hdr.Size = int64(out.Len())
+		tw.WriteHeader(hdr)
+		if l, err := tw.Write(out.Bytes()); err != nil || l != int(hdr.Size) {
+			return fmt.Errorf("error writing %q %q: wrote %v bytes: %v", name, hdr.Name, l, err)
+		}
+
+		atomic.AddInt64(&numFiles, 1)
+		bar.Set("prefix", fmt.Sprintf("Canonicalizing... %d files | ", numFiles))
```
Comment (codyoss) [inline canonical/main.go:305]: Does numFiles need to be loaded, honest question?

**Review by codyoss — COMMENTED** [review]
Looks good overall, just a couple of small things

File: canonical/main.go (hunk 2 above)
Comment (tbpg) [reply canonical/main.go]: Done (`SetAttrSelection`).

**Review by tbpg — COMMENTED** [review]
Thank you!

File: canonical/main.go (hunk 4 above)
Comment (tbpg) [reply canonical/main.go:149]: Done.

File: canonical/main.go (hunk 5 above)
Comment (tbpg) [reply canonical/main.go:157]: Done.

File: canonical/main.go (hunk 7 above)
Comment (tbpg) [reply canonical/main.go:168]: Done.

File: canonical/main.go (hunk 6 above)
Comment (tbpg) [reply canonical/main.go:178]: Removed. Was seeing a panic in some ctrl+c situations, but don't see it now.

File: canonical/main.go (hunk 8 above)
Comment (tbpg) [reply canonical/main.go:211]: Done.

File: canonical/main.go (hunk 9 above)
Comment (tbpg) [reply canonical/main.go:216]: Done.

File: canonical/main.go (hunk 10 above)
Comment (tbpg) [reply canonical/main.go:223]: Done. Just removed the explicit closes.

File: canonical/main.go (hunk 11 above)
Comment (tbpg) [reply canonical/main.go]: I'm not sure what you mean by loaded. But, it's very nice to see the number of files go up as the tool is running. Otherwise, it's not clear if processing is stalled.

In one case, I had a deadlock in the code, didn't have this progress prefix, and didn't know it was stuck until I ctrl+c'd.

**Review by tbpg — COMMENTED** [review]
Thanks!

File: canonical/main.go (hunk 11 above)
Comment (codyoss) [reply canonical/main.go]: Sorry I should have been more specific I meant [atomic.LoadInt64](https://pkg.go.dev/sync/atomic#LoadInt64)

**Review by tritone — APPROVED** [review]

**Review by codyoss — APPROVED** [review]

**Review by fhinkel — APPROVED** [review]
Rubberstamp lgtm
